src/tactic/move_add.lean

```diff
@@ -0,0 +1,250 @@
+local context, then `move_add [f, ← g] at hp₁ hp₂` performs the rearranging at `hp₁` and `hp₂`.
+
+##  Implementation notes
+
+The implementation of `move_add` only move the terms specified by the user (and rearranges
+parentheses).
+
+An earlier version of this tactic also had a relation on `expr` that performed a sorting on the
+terms that were not specified by the user.  This is very easy to implement, if desired, but is not
+part of this tactic.  We had used the order given by the `≤` on `string` and a small support for
+sorting `monomial`s by increasing degree.
+
+Note that the tactic `abel` already implements a very solid heuristic for normalizing terms in an
+additive commutative semigroup and produces expressions in more or less standard form.
+The scope of `move_add` is different: it is designed to make it easy to move individual terms
+around a sum.
+
+##  Future work
+
+* Currently, `_` in user input generate side-goals that Lean should be able to close automatically.
+  Is it possible to avoid this?
+* Using `move_add at *` generates a message mentioning which hypotheses did not change.  It should
+  instead be silent, unless no hypothesis changes.
+* No message is reported if one of the user-provided expressions is not used.
+* Add support for `neg` and additive groups?
+* Add optional different operations than `+`, most notably `*`?
+* Add functionality for moving terms across the two sides of an in/dis/equality.
+  E.g. it might be desirable to have `to_lhs [a]` that converts `b + c = a + d` to `a + b + c = d`.
+* Add more tests.
+-/
+
+open tactic
+
+/-- A `tactic (option expr)` that either finds the first entry `f` of `lc` that unifies with `e`
+and returns `some f` or returns `none`. -/
+meta def expr.find_in (e : expr) (lc : list expr) : tactic (option expr) :=
+do
+  h ← lc.mfilter $ λ e', succeeds $ unify e e',
+  match h with
+  | [] := none
+  | (f::l) := return $ some f
+  end <|>
+return none
+
+/--  Given a list `lp` of `bool × pexpr` and a list `sl` of `expr`, scan the elements of `lp` one
+at a time and produce 3 sublists of `sl`.
+
+If `(tf,pe)` is the first element of `lp`, we look for the first element of `sl` that unifies with
+`pe.to_expr`.  If no such element exists, then we discard `(tf,pe)` and move along.
+If `eu ∈ sl` is the first element of `sl` that unifies with `pe.to_expr`, then we add `eu` as the
+next element of either the first or the second list, depending on the boolean `tf` and we remove
+`eu` from the list `sl`.  In this case, we continue our scanning with the next element of `lp`,
+replacing `sl` by `sl.erase eu`.
+
+Once we exhausts the elements of `lp`, we return the three lists:
+* first the list of elements of `sl` that came from an element of `lp` whose boolean was `tt`,
+* next the ununified elements of `sl` and
+* finally the elements of `sl` that came from an element of `lp` whose boolean was `ff`.
```

Comment:
This looks wrong to me - aren't the second and third entries swapped?

Reply:
I think that this is correct: if you look at the current version of `list.combined`, you can see that it returns 
```lean
(l1 ++ l3 ++ l2)
```
with the "third" list in the middle.  The third list is the one that is left-over from `sl` after all the unification and removal has taken place.

Note: I am about to push a newer version with a fourth entry, corresponding to the list of unused variables.  While the answer still stands, there will be a fourth list in the product!

Reply:
The docstring says "ununified elements of `sl`" appear in the second place. Your comment above says the "third list is the one that is left-over". One of these statements is wrong, and I think it's the docstring.

Reply:
Maybe I have looked at this for too long: the third list is the one labeled `l3` and it is placed as the middle sublist of the output.  I am trying to be careful about the third factor in a product and the middle portion of a list that is obtained by concatenating.  I hope that this clarifies.  Otherwise, I will need to take a break, distract myself for a little bit and then come back and read this again!

Reply:
I am sorry: you are completely right!  This is the step *before* the one that described in the doc-string.  I will update it with the correct version, which is indeed the one where the lists are "unified with left arrow", next "unified without arrow", finally "left-overs in `sl`".

Thanks!  And sorry for taking so long to realize that what you were saying was correct!